cmd/signature-v4-utils.go

```diff
@@ -260,3 +260,28 @@ func signV4TrimAll(input string) string {
 	// unicode.IsSpace() internally here) to one space and return
 	return strings.Join(strings.Fields(input), " ")
 }
+
+// checkMetaHeaders will check if the metadata from header/url is the same with the one from signed headers
+func checkMetaHeaders(signedHeadersMap http.Header, r *http.Request) APIErrorCode {
+	// check values from http header
+	for k, val := range r.Header {
+		if stringsHasPrefixFold(k, "X-Amz-Meta-") {
+			if signedHeadersMap.Get(k) == val[0] {
+				continue
+			}
+			return ErrUnsignedHeaders
+		}
+	}
```

Comment:
only one block is needed `r.Form` has both r.Header and `r.URL.Query()`

Reply:
In the tests add `r.ParseForm()` such that tests would honor `r.Form` properly

Reply:
> only one block is needed `r.Form` has both r.Header and `r.URL.Query()`

`r.Form` contains only values from URL so I cannot check the values for `r.Header` using only `r.Form` and that's why I use two for blocks

Reply:
Oh yes you are right.